TxBuilder order inputs on signing

hewison-chris opened the pull request:
When we generate `Transaction`s it is now required that the `input`s are
ordered by their `utxo` values.
In `TxBuilder` this was not the case and the Faucet was found to be
generating some txs with invalid `input`s in the `Transaction`s.

Fixes #2156

hewison-chris committed:
TxBuilder order inputs on signing

When we generate `Transaction`s it is now required that the `input`s are
ordered by their `utxo` values.
In `TxBuilder` this was not the case and the Faucet was found to be
generating some txs with invalid `input`s in the `Transaction`s.

Geod24 commented:
Why the new variables ?

hewison-chris commented:
> Why the new variables ?

Just because I took the changes from before. But I can remove them now. Will do.

kchulj commented:
> Just because I took the changes from before. But I can remove them now. Will do.

It may be a bit confusing due to the identical name.

hewison-chris commented:
> > Just because I took the changes from before. But I can remove them now. Will do.
> 
> It may be a bit confusing due to the identical name.

There is no longer a need to create temporary variables to hold the arrays pre-sorting as the `Transaction` can be created and the inputs and outputs added unsorted initially. We sort before we finalize and return the transaction.

kchulj commented:
> There is no longer a need to create temporary variables to hold the arrays pre-sorting as the `Transaction` can be created and the inputs and outputs added unsorted initially. We sort before we finalize and return the transaction.

Yes, the change is simple and LGTM.

kchulj commented:
Is there a `unittest` related to this?

hewison-chris commented:
> Is there a `unittest` related to this?

I will add assert in one of the tests in TxBuilder for sorted inputs and outputs.

hewison-chris committed:
Unit test update for TxBuilder

This adds a check that the `input`s and `output`s are sorted by the
`sign` function in `TxBuilder`.

kchulj reviewed (approved):
LGTM

linked0 reviewed (approved):
LGTM

hewison-chris committed:
Unit test update: VariableBlockSize

The unit test has failed sometimes recently in ci. This adds an
assertion if the number of transactions are not as expected.
